handler/src/main/java/io/netty/handler/ssl/ConscryptAlpnSslEngine.java

```diff
@@ -0,0 +1,163 @@
+        return isAvailable() && engineClass.isInstance(engine);
+    }
+
+    static ConscryptAlpnSslEngine newClientEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ClientEngine((OpenSSLEngineImpl) engine, applicationNegotiator);
+    }
+
+    static ConscryptAlpnSslEngine newServerEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ServerEngine((OpenSSLEngineImpl) engine, applicationNegotiator);
+    }
+
+    private ConscryptAlpnSslEngine(OpenSSLEngineImpl engine, List<String> protocols) {
+        super(engine);
+
+        // Register for completion of the handshake.
+        engine.setHandshakeListener(this);
+
+        // Set the list of supported ALPN protocols on the engine.
+        setAlpnProtocols(engine, protocols);
+    }
+
+    OpenSSLEngineImpl getWrappedConscryptEngine() {
+        return (OpenSSLEngineImpl) getWrappedEngine();
+    }
+
+    private static void setAlpnProtocols(OpenSSLEngineImpl engine, List<String> protocols) {
+        engine.setAlpnProtocols(toProtocolArray(protocols));
+    }
+
+    private static String getAlpnSelectedProtocol(OpenSSLEngineImpl engine) {
+        byte[] data = engine.getAlpnSelectedProtocol();
+        return data == null ? null : new String(data, CharsetUtil.US_ASCII);
+    }
+
+    private static byte[] toProtocolArray(List<String> protocols) {
+        int length = 0;
+        for (int i = 0; i < protocols.size(); ++i) {
+            length += 1 + protocols.get(i).length();
+        }
+        byte[] data = new byte[length];
+        for (int dataIndex = 0, i = 0; i < protocols.size(); ++i) {
+            String protocol = protocols.get(i);
+            int protocolLength = protocol.length();
+
+            // Add the length prefix.
+            data[dataIndex++] = (byte) protocolLength;
+            for (int ci = 0; ci < protocolLength; ++ci) {
+                // Assume US-ASCII.
+                data[dataIndex++] = (byte) protocol.charAt(ci);
+            }
+        }
+        return data;
+    }
+
+    private static final class ClientEngine extends ConscryptAlpnSslEngine {
+        private final ProtocolSelectionListener protocolListener;
+
+        ClientEngine(OpenSSLEngineImpl engine,
+                JdkApplicationProtocolNegotiator applicationNegotiator) {
+            super(engine, applicationNegotiator.protocols());
+            protocolListener = checkNotNull(applicationNegotiator
+                            .protocolListenerFactory().newListener(this, applicationNegotiator.protocols()),
+                    "protocolListener");
+        }
+
+        @Override
+        public void onHandshakeFinished() throws SSLException {
+            String protocol = getAlpnSelectedProtocol(getWrappedConscryptEngine());
+            try {
+                protocolListener.selected(protocol);
+            } catch (SSLException e) {
```

Comment:
Unless I'm missing something, I don't think we can do that since we don't want to wrap it in a `RuntimeException`.